job-runner/chain/reaper.go

```diff
@@ -0,0 +1,435 @@
+// Copyright 2018, Square, Inc.
+
+package chain
+
+import (
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/square/spincycle/proto"
+	rm "github.com/square/spincycle/request-manager"
+	"github.com/square/spincycle/retry"
+)
+
+const (
+	// Number of times to attempt sending chain state / SJC to RM.
+	FINALIZE_TRIES = 5
+	// Time to wait between the above tries.
+	FINALIZE_RETRY_WAIT = 500 * time.Millisecond
+)
+
+// A JobReaper handles jobs and chains that have finished running.
+type JobReaper interface {
+	// Reap takes a done job, saves its state, and then prepares the chain for
+	// whatever should be done next.
+	Reap(job proto.Job)
+
+	// Finalize finalizes a job chain, sending the Request Manager its final state
+	// or information that can be used to resume running the chain later.
+	Finalize()
+}
+
+// A ReaperFactory makes new JobReapers.
+type ReaperFactory interface {
+	MakeRunning(chan proto.Job) JobReaper
+	MakeSuspended() JobReaper
+	MakeStopped() JobReaper
+}
+
+type reaperFactory struct {
+	reaper reaper
+}
+
+// Create a new ReaperFactory.
+func NewReaperFactory(chain *chain, chainRepo Repo, rmc rm.Client) ReaperFactory {
+	return NewReaperFactoryWithRetries(chain, chainRepo, rmc, FINALIZE_TRIES, FINALIZE_RETRY_WAIT)
+}
+
+// Create a new ReaperFactory with custom settings for retrying RMClient calls.
+// Used in testing to decrease default retry wait duration.
+func NewReaperFactoryWithRetries(chain *chain, chainRepo Repo, rmc rm.Client, rmTries int, rmRetryWait time.Duration) ReaperFactory {
```

Comment:
I don't see why this is necessary? It strikes me as an anti-pattern: two factories producing the same thing. Difficult to explain/show alternative here. Can discuss in person.